MediaQueryList.matches should update parent document layout for viewport-dependent media queries

rniwa opened the pull request:
#### e66751caa5cc6a8968a262df6c027f19dcc6c0f2
<pre>
MediaQueryList.matches should update parent document layout for viewport-dependent media queries
<a href="https://bugs.webkit.org/show_bug.cgi?id=189583">https://bugs.webkit.org/show_bug.cgi?id=189583</a>

Reviewed by Antti Koivisto.

This patch makes MediaQueryList::matches update the associated document&apos;s owner element&apos;s layout
if the associated media query set contains a viewport dependent expression.

New behavior matches the spec and the behavior of Blink although Blink fails the last test case
in the newly introduced W3C style testharness.js test.

* LayoutTests/fast/css/media-query-matches-in-iframe-expected.txt: Added.
* LayoutTests/fast/css/media-query-matches-in-iframe.html: Added.
* LayoutTests/printing/print-with-media-query-destory-expected.txt: Rebaselined.

* Source/WebCore/css/MediaQueryEvaluator.cpp:
(WebCore::isViewportDependent): Moved to MediaQueryExpression.cpp.
(WebCore::MediaQueryEvaluator::evaluate const):

* Source/WebCore/css/MediaQueryExpression.cpp:
(WebCore::MediaQueryExpression::isViewportDependent const): Moved from MediaQueryEvaluator.cpp.
* Source/WebCore/css/MediaQueryExpression.h:

* Source/WebCore/css/MediaQueryList.cpp:
(WebCore::MediaQueryList::evaluate): This function now dispatches an event on its own instead of
having an out argument for MediaQueryMatcher to use. It now takes MediaQueryMatcher::EventMode
which specifies whether we should be synchronously dispatching a change event or not.
(WebCore::MediaQueryList::matches): Update the layout of the owner element&apos;s document
if the associated media query set contains a viewport dependent expression.
* Source/WebCore/css/MediaQueryList.h:
(WebCore::MediaQueryList::m_needsNotification): Added.

* Source/WebCore/css/MediaQueryMatcher.cpp:
(WebCore::MediaQueryMatcher::evaluateAll): Now takes EventMode which specifies whether we&apos;re
updating the rendering in which case the event should be dispatched now, or we&apos;re updating
for viewport dependent expression in one of the media queries.
* Source/WebCore/css/MediaQueryMatcher.h:

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::evaluateMediaQueriesAndReportChanges):

Canonical link: <a href="https://commits.webkit.org/253123@main">https://commits.webkit.org/253123@main</a>
</pre>

anttijk commented on Source/WebCore/css/MediaQueryList.cpp:
```diff
@@ -104,7 +112,24 @@ void MediaQueryList::setMatches(bool newValue)
+            for (auto& query : m_media->queryVector()) {
+                for (auto& expression : query.expressions())
+                    isViewportDependent = isViewportDependent || expression.isViewportDependent();
+            }
```
You can bail out from the the loops on the first viewport-dependent expression. It might be be good to make this a lambda.

rniwa replied:
Oh, duh. That's a good point.

rniwa replied:
Will do this instead:

```cpp
bool isViewportDependent = [&]() {
    for (auto& query : m_media->queryVector()) {
        for (auto& expression : query.expressions()) {
            if (expression.isViewportDependent())
                return true;
        }
    }
    return false;
}();
```

rniwa committed:
MediaQueryList.matches should update parent document layout for viewport-dependent media queries
https://bugs.webkit.org/show_bug.cgi?id=189583

Reviewed by Antti Koivisto.

This patch makes MediaQueryList::matches update the associated document's owner element's layout
if the associated media query set contains a viewport dependent expression.

New behavior matches the spec and the behavior of Blink although Blink fails the last test case
in the newly introduced W3C style testharness.js test.

* LayoutTests/fast/css/media-query-matches-in-iframe-expected.txt: Added.
* LayoutTests/fast/css/media-query-matches-in-iframe.html: Added.
* LayoutTests/printing/print-with-media-query-destory-expected.txt: Rebaselined.

* Source/WebCore/css/MediaQueryEvaluator.cpp:
(WebCore::isViewportDependent): Moved to MediaQueryExpression.cpp.
(WebCore::MediaQueryEvaluator::evaluate const):

* Source/WebCore/css/MediaQueryExpression.cpp:
(WebCore::MediaQueryExpression::isViewportDependent const): Moved from MediaQueryEvaluator.cpp.
* Source/WebCore/css/MediaQueryExpression.h:

* Source/WebCore/css/MediaQueryList.cpp:
(WebCore::MediaQueryList::evaluate): This function now dispatches an event on its own instead of
having an out argument for MediaQueryMatcher to use. It now takes MediaQueryMatcher::EventMode
which specifies whether we should be synchronously dispatching a change event or not.
(WebCore::MediaQueryList::matches): Update the layout of the owner element's document
if the associated media query set contains a viewport dependent expression.
* Source/WebCore/css/MediaQueryList.h:
(WebCore::MediaQueryList::m_needsNotification): Added.

* Source/WebCore/css/MediaQueryMatcher.cpp:
(WebCore::MediaQueryMatcher::evaluateAll): Now takes EventMode which specifies whether we're
updating the rendering in which case the event should be dispatched now, or we're updating
for viewport dependent expression in one of the media queries.
* Source/WebCore/css/MediaQueryMatcher.h:

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::evaluateMediaQueriesAndReportChanges):

Canonical link: https://commits.webkit.org/253123@main
